### PR TITLE
[CI] gpu test with pytorch nightly

### LINKED ISSUE #2090: [test] Add tests against PT nightly using GPUs
Testing against nightly build.

### PR DESCRIPTION
**Pull request by ydcjeff**
## What does this PR do?

Fixes #2090 
Creates `.drone.jsonnet` for multiple testing

**DO NOT MERGE THIS UNTIL WE SCALE UP DRONE**

### CODE REVIEW

**Review by Borda — COMMENTED** [review]
let's write it as jsonnet script because just one parameter changes, right?
- https://discourse.drone.io/t/porting-matrix-builds-to-1-0-multi-machine-pipelines/2966
- https://github.com/suzuki-shunsuke/drone-jsonnet-generator

**Comment by ydcjeff** [comment]
> let's write it as jsonnet script because just one parameter changes, right?
> 
>     * https://discourse.drone.io/t/porting-matrix-builds-to-1-0-multi-machine-pipelines/2966
> 
>     * https://github.com/suzuki-shunsuke/drone-jsonnet-generator

Yes, only docker image changes.
Okay, will take a look

**Comment by Borda** [comment]
I think that then we would need to coordinate, because when you add it, ping me and would need to a temporary change the Drone config file, so lets keep the old one and add this jsonnet as a new file...

**Comment by Borda** [comment]
@ydcjeff we can move forward...

**Comment by ydcjeff** [comment]
@Borda btw, are we going to run them concurrently or in parallel?

**Comment by Borda** [comment]
well mi mind wait with this one as we are reaching some computational limits on Drone, sometimes. now we have a queue for about 2hours and with running this one it will be even double... so let's prepare it but wait with merging it...

**Comment by ydcjeff** [comment]
Yea, I am fine with anyway. I have now setup drone to run one after another in jsonnet.
Here's is jsonnet to yaml output.

```yaml
---
kind: pipeline
type: docker
name: torch-GPU

platform:
  os: linux
  arch: amd64

steps:
- name: testing
  image: pytorchlightning/pytorch_lightning:base-cuda-py3.7-torch1.6
  commands:
  - export PATH=$PATH:/root/.local/bin
  - python --version
  - pip install pip -U
  - pip --version
  - nvidia-smi
  - apt-get update && apt-get install -y cmake
  - pip install -r ./requirements/base.txt -q --upgrade-strategy only-if-needed
  - pip install -r ./requirements/devel.txt -q --upgrade-strategy only-if-needed
  - pip install -r ./requirements/examples.txt -q --upgrade-strategy only-if-needed
  - pip list
  - python -c 'import torch ; print(' & '.join([torch.cuda.get_device_name(i) for i in range(torch.cuda.device_count())]) if torch.cuda.is_available() else 'only CPU')'
  - coverage run --source pytorch_lightning -m py.test pytorch_lightning tests -v --durations=25
  - python -m py.test benchmarks pl_examples -v --maxfail=2 --durations=0
  - coverage report
  - codecov --token $CODECOV_TOKEN --flags=gpu,pytest --name='GPU-coverage' --env=linux --build $DRONE_BUILD_NUMBER --commit $DRONE_COMMIT
  - python tests/collect_env_details.py
  environment:
    CODECOV_TOKEN:
      from_secret: codecov_token
    HOROVOD_GPU_ALLREDUCE: NCCL
    HOROVOD_GPU_BROADCAST: NCCL
    HOROVOD_WITHOUT_MPI: 1
    HOROVOD_WITHOUT_MXNET: 1
    HOROVOD_WITHOUT_TENSORFLOW: 1
    HOROVOD_WITH_GLOO: 1
    HOROVOD_WITH_PYTORCH: 1
    MKL_THREADING_LAYER: GNU
    SLURM_LOCALID: 0

trigger:
  branch:
  - master
  event:
  - push
  - pull_request

---
kind: pipeline
type: docker
name: torch-GPU-nightly

platform:
  os: linux
  arch: amd64

steps:
- name: testing
  image: pytorchlightning/pytorch_lightning:base-cuda-py3.7-torch1.7
  commands:
  - export PATH=$PATH:/root/.local/bin
  - python --version
  - pip install pip -U
  - pip --version
  - nvidia-smi
  - apt-get update && apt-get install -y cmake
  - pip install -r ./requirements/base.txt -q --upgrade-strategy only-if-needed
  - pip install -r ./requirements/devel.txt -q --upgrade-strategy only-if-needed
  - pip install -r ./requirements/examples.txt -q --upgrade-strategy only-if-needed
  - pip list
  - python -c 'import torch ; print(' & '.join([torch.cuda.get_device_name(i) for i in range(torch.cuda.device_count())]) if torch.cuda.is_available() else 'only CPU')'
  - coverage run --source pytorch_lightning -m py.test pytorch_lightning tests -v --durations=25
  - python -m py.test benchmarks pl_examples -v --maxfail=2 --durations=0
  - coverage report
  - codecov --token $CODECOV_TOKEN --flags=gpu,pytest --name='GPU-coverage' --env=linux --build $DRONE_BUILD_NUMBER --commit $DRONE_COMMIT
  - python tests/collect_env_details.py
  environment:
    CODECOV_TOKEN:
      from_secret: codecov_token
    HOROVOD_GPU_ALLREDUCE: NCCL
    HOROVOD_GPU_BROADCAST: NCCL
    HOROVOD_WITHOUT_MPI: 1
    HOROVOD_WITHOUT_MXNET: 1
    HOROVOD_WITHOUT_TENSORFLOW: 1
    HOROVOD_WITH_GLOO: 1
    HOROVOD_WITH_PYTORCH: 1
    MKL_THREADING_LAYER: GNU
    SLURM_LOCALID: 0

trigger:
  branch:
  - master
  event:
  - push
  - pull_request

depends_on:
- torch-GPU

...

```

**Comment by ydcjeff** [comment]
> well mi mind wait with this one as we are reaching some computational limits on Drone, sometimes. now we have a queue for about 2hours and with running this one it will be even double... so let's prepare it but wait with merging it...

@Borda regarding with testing nightly, would it be better to start support PT nightly version ahead of one month before stable launch?
also draft those nightly PRs and test them without merging not to do the follow for fixing. I thought this could save some GPU resources.

For now, there were already 2 PRs for PyTorch nightly, started thinking about that...

**Review by Borda — APPROVED** [review]

**Comment by edenlightning** [comment]
hey @ydcjeff hows it going? any way we can help?

**Comment by Borda** [comment]
@edenlightning we can merge it as it now, but to activate it (switch source in Drone config) we need to get scalable Drone testing running because this basically double the number of performed tests and already now we are full...

**Review by awaelchli — APPROVED** [review]

File: .drone.jsonnet [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+        "CODECOV_TOKEN": {
+          from_secret: "codecov_token"
+        },
+        "MKL_THREADING_LAYER": "GNU",
+        "HOROVOD_GPU_OPERATIONS": "NCCL",
+        "HOROVOD_WITH_PYTORCH": 1,
+        "HOROVOD_WITHOUT_TENSORFLOW": 1,
+        "HOROVOD_WITHOUT_MXNET": 1,
+        "HOROVOD_WITH_GLOO": 1,
+        "HOROVOD_WITHOUT_MPI": 1,
```
Comment (Borda) [inline .drone.jsonnet:38]:
```suggestion
```

File: .drone.jsonnet [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+        "python -c 'import torch ; print(' & '.join([torch.cuda.get_device_name(i) for i in range(torch.cuda.device_count())]) if torch.cuda.is_available() else 'only CPU')'",
+        "coverage run --source pytorch_lightning -m py.test pytorch_lightning tests -v --durations=25", // --flake8",
+        "python -m py.test benchmarks pl_examples -v --maxfail=2 --durations=0", // --flake8",
+        // "cd docs; make doctest; make coverage",
+        "coverage report",
+        // see: https://docs.codecov.io/docs/merging-reports
+        "codecov --token $CODECOV_TOKEN --flags=gpu,pytest --name='GPU-coverage' --env=linux --build $DRONE_BUILD_NUMBER --commit $DRONE_COMMIT",
+        // "--build $DRONE_BUILD_NUMBER --branch $DRONE_BRANCH --commit $DRONE_COMMIT --tag $DRONE_TAG --pr $DRONE_PULL_REQUEST",
+        // "- codecov --token $CODECOV_TOKEN --flags=gpu,pytest --build $DRONE_BUILD_NUMBER",
+        "python tests/collect_env_details.py"
```
Comment (Borda) [inline .drone.jsonnet:61]:
```suggestion
        "python --version",
        "pip --version",
        "nvidia-smi",
        "pip install -r ./requirements/devel.txt --upgrade-strategy only-if-needed -v --no-cache-dir",
        "pip list",
        "coverage run --source pytorch_lightning -m pytest pytorch_lightning tests -v --color=yes --durations=25",
        "python -m pytest benchmarks pl_examples -v --color=yes --maxfail=2 --durations=0",
        "coverage report",
        "codecov --token $CODECOV_TOKEN --flags=gpu,pytest --name='GPU-coverage' --env=linux --build $DRONE_BUILD_NUMBER --commit $DRONE_COMMIT",
        "python tests/collect_env_details.py"
```

File: .drone.jsonnet [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+      ],
+    },
+  ],
+  trigger: {
+    branch: "master",
```
Comment (Borda) [inline .drone.jsonnet:66]:
```suggestion
    branch: [
        "master",
        "release/*"
    ],
```

**Review by Borda — COMMENTED** [review]
update after #3658

File: .drone.jsonnet [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+        "coverage report",
+        "codecov --token $CODECOV_TOKEN --flags=gpu,pytest --name='GPU-coverage' --env=linux --build $DRONE_BUILD_NUMBER --commit $DRONE_COMMIT",
+        "python tests/collect_env_details.py"
+      ],
+    },
+  ],
+  trigger: {
+    branch: [
+        "master",
+        "release/*"
```
Comment (Borda) [inline .drone.jsonnet:51]:
```suggestion
      "master",
      "release/*"
```

File: .drone.jsonnet [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+local pipeline(name, image) = {
+  kind: "pipeline",
+  type: "docker",
+  name: name,
+  steps: [
+    {
+      name: "testing",
+      image: image,
+      environment: {
+        "SLURM_LOCALID": 0,
```
Comment (Borda) [inline .drone.jsonnet:28]:
```suggestion
```

**Comment by Borda** [comment]
This is not tested, but still, I would merge it as using it as a baseline and debug it later...
For testing, we need a temporary switch Drone setting and see if it works...

**Comment by ydcjeff** [comment]
Okay 👌

File: .drone.jsonnet [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+        },
+        "MKL_THREADING_LAYER": "GNU",
+      },
+      commands: [
+        "python --version",
+        "pip --version",
+        "nvidia-smi",
+        "pip install -r ./requirements/devel.txt --upgrade-strategy only-if-needed -v --no-cache-dir",
+        "pip list",
+        "coverage run --source pytorch_lightning -m pytest pytorch_lightning tests -v --color=yes --durations=25",
```
Comment (ydcjeff) [inline .drone.jsonnet:39]:
```suggestion
        "coverage run --source pytorch_lightning -m pytest pytorch_lightning tests -v -ra --color=yes --durations=25",
```

File: .drone.jsonnet [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+        "python -m pytest benchmarks pl_examples -v --color=yes --maxfail=2 --durations=0",
```
Comment (ydcjeff) [inline .drone.jsonnet:40]:
```suggestion
        "python -m pytest benchmarks pl_examples -v -ra --color=yes --maxfail=2 --durations=0",
```

**Comment by ydcjeff** [comment]
I messed up with my forked repo's master and re-forked again.
Let me know if you want to merge, I can create a new one.

**Review by tchaton — APPROVED** [review]

**Review by SeanNaren — APPROVED** [review]

**Comment by Borda** [comment]
@ydcjeff seems this is not valid as you lost the original fork, mind create it again and refer to this already approved PR 🐰

**Pull request closed without merging**